### PR TITLE
[FLINK-13121] [table-planner-blink] Set batch properties to runtime in blink batch executor

### PR DESCRIPTION
**Pull request by XuPingyong**
## What is the purpose of the change

1. set batch properties to StreamGraph
2. set managedMemory size to ResourceSpec
3. set shuffleMode to PartitionTransformations.

## Brief change log

1. set batch properties to StreamGraph
2. set managedMemory size to ResourceSpec
3. set shuffleMode to PartitionTransformations.

## Verifying this change

UT

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: ( no)
  - The runtime per-record code paths (performance sensitive): (no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 1]
```diff
@@ -42,47 +42,105 @@
...
-	public static StreamGraph generateStreamGraph(
-		StreamExecutionEnvironment execEnv,
-		List<Transformation<?>> transformations,
-		String jobName) throws Exception {
-		// TODO avoid cloning ExecutionConfig
-		ExecutionConfig executionConfig = InstantiationUtil.clone(execEnv.getConfig());
+	/**
+	 * Backup previous streamEnv config and set batch configs.
+	 */
+	private void backupAndUpdateStreamEnv(StreamExecutionEnvironment execEnv) {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Why should we backup old stream env and restore it after job execution? I think this kind of logic definitely needs some unit tests.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala [hunk 2]
```diff
@@ -116,25 +75,40 @@ class BatchExecExchange(
 
   override def explainTerms(pw: RelWriter): RelWriter = {
     super.explainTerms(pw)
-      .itemIf("exchange_mode", requiredExchangeMode.orNull,
-        requiredExchangeMode.contains(DataExchangeMode.BATCH))
+      .itemIf("shuffle_mode", requiredShuffleMode.orNull,
+        requiredShuffleMode.contains(ShuffleMode.BATCH))
+        .itemIf("full_dam_behavior", getDamBehavior,
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: why we add this?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/ExecutionConfigOptions.java [hunk 3]
```diff
@@ -195,4 +195,11 @@
 							"means a kind of disabled operator. Its default value is empty that means no operators are disabled. " +
 							"If the configure's value is \"NestedLoopJoin, ShuffleHashJoin\", NestedLoopJoin and ShuffleHashJoin " +
 							"are disabled. If configure's value is \"HashJoin\", ShuffleHashJoin and BroadcastHashJoin are disabled.");
+
+	public static final ConfigOption<Boolean> SQL_EXEC_SHUFFLE_MODE_ALL_BATCH =
+			key("sql.exec.shuffle-mode.all-batch")
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/ExecutionConfigOptions.java]: change to sql.exec.shuffle-mode?
value could be "batch", "pipeline" and so on

Comment (XuPingyong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/api/ExecutionConfigOptions.java]: ok, changed

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala (hunk 2 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: had removed it...

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 1 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: As batch job execution can change the properties of streamEnv to affect its reuse.  Added UT.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 4]
```diff
@@ -42,47 +43,108 @@
 @Internal
 public class BatchExecutor extends ExecutorBase {
 
+	private boolean enableObjectReuse;
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: create a structure for these backed up variables?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 5]
```diff
@@ -42,47 +43,108 @@
...
+				streamGraph.setBlockingConnectionsBetweenChains(true);
+			}
+		} finally {
+			restoreStreamEnv(execEnv);
+		}
+		return streamGraph;
+	}
+
+	private boolean isShuffleModeAllBatch() {
+		ExecutionConfig.GlobalJobParameters parameters = getExecutionEnvironment().getConfig().getGlobalJobParameters();
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Why don't you read from `TableConfig`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 6]
```diff
@@ -42,47 +43,108 @@
...
+			executionConfig.setDefaultInputDependencyConstraint(inputDependencyConstraint);
+		}
+	}
+
+	/**
+	 * Translates transformationList to streamGraph.
+	 */
+	public StreamGraph generateStreamGraph(
+			List<Transformation<?>> transformations,
+			String jobName) throws Exception {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: This function will not throw Exception

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 7]
```diff
@@ -42,47 +43,108 @@
...
+		StreamExecutionEnvironment execEnv = getExecutionEnvironment();
+		backupAndUpdateStreamEnv(execEnv);
+		transformations.forEach(execEnv::addOperator);
+		StreamGraph streamGraph;
+		try {
+			streamGraph = execEnv.getStreamGraph(getNonEmptyJobName(jobName));
+			// If one transformation uses managed memory, all transformations should set managed memory to 0.
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: What does this mean?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java [hunk 8]
```diff
@@ -42,47 +43,108 @@
...
+
+	/**
+	 * Translates transformationList to streamGraph.
+	 */
+	public StreamGraph generateStreamGraph(
+			List<Transformation<?>> transformations,
+			String jobName) throws Exception {
+		StreamExecutionEnvironment execEnv = getExecutionEnvironment();
+		backupAndUpdateStreamEnv(execEnv);
+		transformations.forEach(execEnv::addOperator);
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: If Exception is thrown here, you will not restore execEnv

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala [hunk 9]
```diff
@@ -116,25 +75,39 @@ class BatchExecExchange(
...
-      case Some(mode) if mode eq DataExchangeMode.BATCH => mode
-      case _ => DataExchangeMode.PIPELINED
+  private[flink] def getShuffleMode(tableConf: Configuration): ShuffleMode = {
+    requiredShuffleMode match {
+      case Some(mode) if mode eq ShuffleMode.BATCH => mode
+      case _ => getShuffleModeFromConfig(tableConf)
+    }
+  }
+
+  private def getShuffleModeFromConfig(tableConf: Configuration): ShuffleMode = {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: This method name is almost the same with previous one, it's confusing

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 4 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Created.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 6 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Removed it.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 7 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Updated it.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala (hunk 9 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecExchange.scala]: fixed

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 8 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Now I restore execEnv at last and remove throwing exception.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java (hunk 5 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/executor/BatchExecutor.java]: Fixed it.

**Comment by KurtYoung** [comment]
travis failed on irrelevant case: https://travis-ci.org/KurtYoung/flink/builds/557582152

**Pull request closed without merging**